[DTensor][BE] improve DTensor ops correctness check utils

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158112
* #158051

**Summary**
Implemented the test pattern described in https://github.com/pytorch/pytorch/pull/157991#discussion_r2196363170 as a util method in `DTensorTestBase`. The difference to `DTensorTestBase._test_op` is:
1. allowing users to specify the `Partial` placement.
2. supporting tree-like output structure.

**Test**
so far only adopt `DTensorTestBase._test_op_on_dtensor` in `DistTensorOpsTest.test_split_on_partial`.
`pytest test/distributed/tensor/test_tensor_ops.py -s -k test_split_on_partial`

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @tianyu-l

XilunWu committed:
[DTensor][BE] imporve DTensor ops correctness check utils

[ghstack-poisoned]

zpcore commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -403,6 +404,21 @@ def setUp(self) -> None:
         super().setUp()
         self._spawn_processes()
 
+    def _test_op_on_dtensor(self, dtensor: DTensor, op_call) -> None:
```
This is super useful! Not necessary in this PR, is it possible we extend arg dtensor into *args, so that we can apply this function to different ops?

XilunWu replied:
yes, or you can just pass in a lambda (see the use case in `test_tensor_ops` I add in this PR).

zpcore reviewed (approved):
LGTM!

wconstab commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -403,6 +404,21 @@ def setUp(self) -> None:
+        this function takes in DTensor object directly as argument and test the equality
+        of calling op on full_tensor() and DTensor.
+
+        Note: for in-place ops, the function may not work as expected if the DTensor
```
shouldn't we be able to fix this limitation easily by just cloning the full tensor?

XilunWu replied:
I'm not very sure that this is an actual problem. But I guess you're right, I should just preventively do `full_tensor().detach().copy()`

Skylion007 commented:
Typo on git commit message though

XilunWu committed:
Update on "[DTensor][BE] improve DTensor ops correctness check utils"


**Summary**
Implemented the test pattern described in https://github.com/pytorch/pytorch/pull/157991#discussion_r2196363170 as a util method in `DTensorTestBase`. The difference to `DTensorTestBase._test_op` is:
1. allowing users to specify the `Partial` placement.
2. supporting tree-like output structure.

**Test**
so far only adopt `DTensorTestBase._test_op_on_dtensor` in `DistTensorOpsTest.test_split_on_partial`.
`pytest test/distributed/tensor/test_tensor_ops.py -s -k test_split_on_partial`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k tianyu-l

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge
